### PR TITLE
Add image comparision tests for `VideoAsset` and `VideoStream`

### PR DESCRIPTION
**Pull request by Wumpf**
Commit by commit!

To enable testing video, we had to to add a store hub to the test harness. This made a few things a bit more complex but ultimiatively simplifes aspects of the harness!

The new image comparision test for videos makes sure to tests a few corner cases by seeking within the video (see code for details).

Furthermore, this fixes a small bug in `VideoStream` where we'd incorrectly show an error when seeking before the video.

* [ ] pass full ci-check (changes in video asset may affect snippet comparision)
* [ ] @lucasmerlin does the usage of testharness in `video.rs` make sense here?
* [ ] @abey79 does the changes to the `TestContext` make sense

### CODE REVIEW

**Comment by Wumpf** [comment]
@rerun-bot full-check

**Comment by Wumpf** [comment]
@rerun-bot full-check

**Comment by Wumpf** [comment]
@rerun-bot full-check

**Comment by Wumpf** [comment]
@rerun-bot full-check

maybe this time I got good threshholds?

**Comment by Wumpf** [comment]
windows is spuriously failing to decode altogether :/

**Comment by Wumpf** [comment]
@rerun-bot full-check


tried to make it more robust, let's see if that helps (hard to tell though if we just gonna get lucky)

**Comment by Wumpf** [comment]
@rerun-bot full-check

**Comment by Wumpf** [comment]
@rerun-bot full-check

File: crates/viewer/re_viewport/src/test_context_ext.rs [hunk 1]
```diff
@@ -56,16 +56,17 @@ impl TestContextExt for TestContext {
...
                     viewport_blueprint.save_to_blueprint_store(ctx);
                 });
 
                 self.handle_system_commands();
 
                 // Reload the blueprint store and execute all view queries.
+                let blueprint_query = self.blueprint_query.clone();
                 let viewport_blueprint =
-                    ViewportBlueprint::from_db(&self.blueprint_store, &self.blueprint_query);
+                    ViewportBlueprint::from_db(self.active_blueprint(), &blueprint_query);
```
Comment (abey79) [inline crates/viewer/re_viewport/src/test_context_ext.rs:69]: it feels slightly off that the blueprint would be accessed by two different means here.

File: crates/viewer/re_view_spatial/src/visualizers/video/video_frame_reference.rs [hunk 2]
```diff
@@ -212,6 +212,9 @@ impl VideoFrameReferenceVisualizer {
                             if let Some([w, h]) = video.dimensions() {
                                 video_resolution = glam::vec2(w as _, h as _);
                             }
+                            if err.should_rerequest_frame() {
```
Comment (abey79) [inline crates/viewer/re_view_spatial/src/visualizers/video/video_frame_reference.rs]: nit: s/rerequest/request_repaint/

**Review by abey79 — APPROVED** [review]
I primarily focused on the `TestContext` changes and they lgtm. Let's hope we can be proud of `StoreHub` again by end of summer 😅

File: crates/viewer/re_viewport/src/test_context_ext.rs (hunk 1 above)
Comment (Wumpf) [reply crates/viewer/re_viewport/src/test_context_ext.rs:69]: can't access the harness in `run` and don't have the ctx outside :shrug:

Comment (abey79) [reply crates/viewer/re_viewport/src/test_context_ext.rs:69]: ah sorry, I overlooked that fact that the first one is in a closure. nvm then.

**Comment by Wumpf** [comment]
@rerun-bot full-check

File: tests/assets/video/Big_Buck_Bunny_1080_1s_av1.mp4 [hunk 3]
```diff
@@ -0,0 +1,3 @@
+version https://git-lfs.github.com/spec/v1
```
Comment (emilk) [inline tests/assets/video/Big_Buck_Bunny_1080_1s_av1.mp4:1]: We should rename `.github/workflows/png_only_on_lfs.yml` and make sure it also checks for mp4

File: crates/store/re_types/src/archetypes/asset_video_ext.rs [hunk 4]
```diff
@@ -39,6 +39,24 @@ impl AssetVideo {
...
+            .as_any()
+            .downcast_ref::<arrow::array::ListArray>()
+            .expect("Video blob data is not a ListArray");
+        let blob_data = blob_list_array
+            .values()
+            .as_any()
+            .downcast_ref::<arrow::array::PrimitiveArray<arrow::datatypes::UInt8Type>>()
+            .expect("Video blob data is not a PrimitiveArray<UInt8>");
+
+        Some(blob_data.values().inner().as_slice())
```
Comment (emilk) [inline crates/store/re_types/src/archetypes/asset_video_ext.rs]: Wouldn't most of this code make more sense in `blob_ext.rs`? So that:

A) it is closer to the logic that controls how it is encoded
B) it can be reused by other archetypes that also has `Blob` in them

File: crates/store/re_types/src/archetypes/asset_video_ext.rs [hunk 5]
```diff
@@ -39,6 +39,24 @@ impl AssetVideo {
...
 
+    /// Returns the bytes of the video blob.
+    pub fn video_blob_bytes(&self) -> Option<&[u8]> {
+        let blob_list_array = self
+            .blob
+            .as_ref()?
+            .array
+            .as_any()
+            .downcast_ref::<arrow::array::ListArray>()
+            .expect("Video blob data is not a ListArray");
```
Comment (emilk) [inline crates/store/re_types/src/archetypes/asset_video_ext.rs]: Let's return `None` instead of crashing

File: crates/store/re_types/src/archetypes/asset_video_ext.rs [hunk 6]
```diff
@@ -39,6 +39,24 @@ impl AssetVideo {
...
+        let blob_data = blob_list_array
+            .values()
+            .as_any()
+            .downcast_ref::<arrow::array::PrimitiveArray<arrow::datatypes::UInt8Type>>()
+            .expect("Video blob data is not a PrimitiveArray<UInt8>");
```
Comment (emilk) [inline crates/store/re_types/src/archetypes/asset_video_ext.rs]: Here too

File: crates/utils/re_video/src/decode/ffmpeg_h264/ffmpeg_libloading.rs [hunk 7]
```diff
@@ -0,0 +1,1155 @@
+//! Send video data to `ffmpeg` over CLI to decode it.
```
Comment (emilk) [inline crates/utils/re_video/src/decode/ffmpeg_h264/ffmpeg_libloading.rs]: Is this whole file a copy of `crates/utils/re_video/src/decode/ffmpeg_h264/ffmpeg.rs`? Did you mean to rename it but created a copy instead?

**Review by emilk — COMMENTED** [review]
Took a quick look; need to eat something before I continue

File: crates/viewer/re_view_spatial/tests/video.rs [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+        ".pixi/envs/default/Library/bin/ffmpeg"
+    })
+}
+
+fn video_test_file_mp4(codec: VideoCodec, need_dts_equal_pts: bool) -> std::path::PathBuf {
+    let codec_str = match codec {
+        VideoCodec::H264 => "h264",
+        VideoCodec::H265 => "h265",
+        VideoCodec::VP9 => "vp9",
+        VideoCodec::VP8 => panic!("Don't have test data for vp8. Mp4 doesn't support it."),
```
Comment (emilk) [inline crates/viewer/re_view_spatial/tests/video.rs]:
```suggestion
        VideoCodec::VP8 => panic!("We don't have test data for vp8, because Mp4 doesn't support vp8."),
```

File: crates/store/re_types/src/archetypes/asset_video_ext.rs (hunk 4 above)
Comment (Wumpf) [reply crates/store/re_types/src/archetypes/asset_video_ext.rs]: huh, yeah you're right. Let's do that!

File: crates/utils/re_video/src/decode/ffmpeg_h264/ffmpeg_libloading.rs (hunk 7 above)
Comment (Wumpf) [reply crates/utils/re_video/src/decode/ffmpeg_h264/ffmpeg_libloading.rs]: argh. When I kept poking things on the weekend I accidentally put in this discarded experimentation file. This is not supposed to be here!

File: crates/viewer/re_view_spatial/tests/video.rs [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+        VideoCodec::AV1 => 100,
+        _ => SnapshotOptions::default().failed_pixel_count_threshold,
+    }
+}
+
+fn test_video(video_type: VideoType, codec: VideoCodec) {
+    let mut test_context = TestContext::new_with_view_class::<re_view_spatial::SpatialView2D>();
+
+    // Use pixi ffmpeg install if available.
+    if cfg!(target_os = "windows") {
```
Comment (emilk) [inline crates/viewer/re_view_spatial/tests/video.rs]: Why only on Windows though?

File: crates/viewer/re_view_spatial/tests/video.rs [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+            assert!(
+                video_data_description
+                    .samples_statistics
+                    .dts_always_equal_pts,
+                "TODO(#10090): Video stream doesn't support bframes"
+            );
+
+            for sample in video_data_description.samples.iter() {
+                let raw_sample_bytes =
+                    &blob_bytes[sample.byte_span.start as usize..sample.byte_span.end() as usize];
```
Comment (emilk) [inline crates/viewer/re_view_spatial/tests/video.rs]: Shorter:

```suggestion
                    &blob_bytes[sample.byte_span.range_usize()];
```

File: crates/store/re_types/src/archetypes/asset_video_ext.rs (hunk 4 above)
Comment (Wumpf) [reply crates/store/re_types/src/archetypes/asset_video_ext.rs]: the only hurdle is that we don't have `Blob` type here. But we can just make one up on the spot 🤷

Comment (Wumpf) [reply crates/store/re_types/src/archetypes/asset_video_ext.rs]: right, there's no blob archetype.
We can have a static method on the blob datatype to help with this kind of thing, but we're dealing with serialized blob data here whereas `Blob` component and `Blob` datatype are about unseralized data

Comment (Wumpf) [reply crates/store/re_types/src/archetypes/asset_video_ext.rs]: Yeah that looks alright :) [4950c87](https://github.com/rerun-io/rerun/pull/10522/commits/4950c87ecd77f113304a4d9b4804ace0adbd8e56)

File: crates/viewer/re_view_spatial/tests/video.rs [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+                });
+            }
+        }
+    }
+
+    let view_id = test_context.setup_viewport_blueprint(|_ctx, blueprint| {
+        blueprint.add_view_at_root(ViewBlueprint::new_with_root_wildcard(
+            re_view_spatial::SpatialView2D::identifier(),
+        ))
+    });
```
Comment (emilk) [inline crates/viewer/re_view_spatial/tests/video.rs:289]:
```suggestion
    let step_dt_seconds = 1.0 / 4.0; // This is also the current egui_kittest default, but let's be explicit since we use `try_run_realtime`.
```

File: crates/viewer/re_view_spatial/tests/video.rs [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+        .with_step_dt(step_dt_seconds)
+        .with_max_steps((max_total_time_seconds / step_dt_seconds) as u64)
+        .with_size(egui::vec2(300.0, 200.0))
+        .build(|ctx| {
+            {
+                let mut time_ctrl = test_context.recording_config.time_ctrl.write();
+                time_ctrl.set_time(TimeInt::from_nanos(
+                    NonMinI64::new(desired_seek_ns.get()).unwrap(),
+                ));
+            }
```
Comment (emilk) [inline crates/viewer/re_view_spatial/tests/video.rs:310]: What is this for?

File: crates/viewer/re_view_spatial/tests/video.rs (hunk 9 above)
Comment (Wumpf) [reply crates/viewer/re_view_spatial/tests/video.rs]: oh no. why did I do that. must have mixed something up when I `pixi_ffmpeg_path` still inline :facepalm:
wondering now if that changes the high threshholds again a bit. I guess on to another round of trying to use lower ones (well, empirically av1 already needs a high threshhold as well though)

File: crates/viewer/re_view_spatial/tests/video.rs (hunk 12 above)
Comment (Wumpf) [reply crates/viewer/re_view_spatial/tests/video.rs:310]: we want several frames to seek to. Since we can't access `test_context` anymore as it's borrowed by `harness` now, we have to do this inside of the method the harness runs

**Comment by Wumpf** [comment]
@rerun-bot full-check

one last time 🤞
